### DISCUSSION TITLE
How can I use postgres string arrays?

### DISCUSSION THREAD

**Discussion by tasn** [SeaQL/sea-orm#487] [Q&A]
Can't find anything that references them in the code. How do I load them? Query them?

**Comment by billy1624** [comment]
Hey @tasn, you mean PostgreSQL array in general? It's still work in progress, check
- #151
- SeaQL/sea-query#225
- SeaQL/sea-query#205

Btw... are you trying to get an array of strings from the db? For example, loading it into a Model like
```rust
pub struct Model {
    pub strings: Vec<String>,
    ...
}
```

**Reply by emmiegit** [reply]
Is there something that can be done in the meantime? I know it generates `#[sea_orm(column_type = "Custom(\"array\".to_owned())")]`, but I'm not sure how to handle that. (It seems to produce a string, and I don't know what format it wants it in)

**Reply by tasn** [reply]
Yeah, postgres arrays in general. Both loading/saving, and querying.

Thanks for the info, so I take that this is a no. Is there a way to use sqlx directly for this? Because IIRC the pool is sea-orm only.

**Reply by billy1624** [reply]
> Thanks for the info, so I take that this is a no.

It's a no for now.

> Is there a way to use sqlx directly for this? Because IIRC the pool is sea-orm only.

I believe you can, just bind a `Vec<T>` or `&[T]` directly.

```rust
// provides `try_next`
use futures::TryStreamExt;

let mut rows = sqlx::query("SELECT * FROM users WHERE email = ?")
    .bind(email)
    .fetch(&mut conn);

while let Some(row) = rows.try_next().await? {
    // map the row into a user-defined domain type
    let email: &str = row.try_get("email")?;
}
```

**Reply by tasn** [reply]
Sorry, I wasn't clear, I meant: is there a way to get the sqlx connection pool from the sea orm pool so that I can actually use sqlx? It looked like the answer may be a no, but I'm not sure.

**Reply by emmiegit** [reply]
Or a way to invoke custom queries using Sea-ORM so we can read/write arrays manually that way?

**Reply by billy1624** [reply]
> is there a way to get the sqlx connection pool from the sea orm pool so that I can actually use sqlx?

No, there is no way of getting it out under existing public API. As we don't want to leak SQLx out to the public.

You can query raw SQL, but I think it would be difficult for you to bind array of values
- https://www.sea-ql.org/SeaORM/docs/basic-crud/raw-sql/#use-raw-query--execute-interface

**Reply by tasn** [reply]
@billy1624, that's my point. So it's not actually possible without creating an additional sqlx connection pool?

**Reply by billy1624** [reply]
> So it's not actually possible without creating an additional sqlx connection pool?

At the moment, yes.
We might support PostgreSQL arrays in the future so won't need to create an additional SQLx connection to handle it.

**Reply by tasn** [reply]
I guess I have a related question: can seaql/sea-orm query into json objects? E.g. can I just store my arrays in jsonb (not optimal, I know) and then query there as a hack until this is resolved?

Thanks!

**Reply by billy1624** [reply]
Yes, you can. Both json & jsonb are supported.
See https://github.com/SeaQL/sea-orm/pull/488/files#diff-7187ac044721cf235077a238354705e993695449040211b2267435dda79f0b62

**Reply by tasn** [reply]
I know jsonb is supported (I'm using it), I'm asking whether there's a way to query into them, so filter by a field so it can replace an array for us. So e.g. I can do:
```
        Self::find().filter(Column::SomeJsonFiled.in(a_value_in_the_array))
```

**Reply by billy1624** [reply]
I'm not familiar with PostgreSQL's json datatype. Can you show me a raw SQL that will perform the intended select query?

**Reply by tasn** [reply]
I'm not too familiar with it either, I was suggesting it as a hack around the lack of PG array support. I haven't queried into it before.

**Reply by tasn** [reply]
Here is how it's done.
```
select * from (SELECT *, ARRAY(SELECT jsonb_array_elements_text(json_with_array)) AS "txt_arr" FROM   endpoint) as endp where 'some_value' = ANY("txt_arr");
```
As you can see, it requires a subquery and it's really ugly.

I think going with PG arrays is the way to go, any ETA on when that would be ready? Even if only to insert/fetch and not necessarily filter?
Because if that's done I can just use `Expr::cust_with_values` to do the actual fetching.

**Reply by tasn** [reply]
OK, nvm, I just figured out this second, after hours of searching, there's an easy way to do it!

https://www.postgresql.org/docs/9.4/functions-json.html#FUNCTIONS-JSONB-OP-TABLE

**Reply by tasn** [reply]
I'm stuck again though as I can't figure out how to serialize arrays so I can match against them.
So if I want to do:
```
WHERE col in ('val1', 'val2')
```

Edit: I mean, I can just generate the string myself, but it's... :|

Edit edit: passing `some_value ??| array[ ?, ? ]` doesn't work. It doesn't evaluate the `?` inside of the array initializer.

**Reply by billy1624** [reply]
Hey @tasn, sorry, I'm confused. Could you please explain more? The difficulties you face.

**Reply by tasn** [reply]
Sorry, I was in the midst of battling things. :)

So I tried doing this which didn't work:
```rust
        let cond = "some_jsonb_array ??| array [ ?, ? ]";
        query = query.filter(Expr::cust_with_values(&cond, vec![str1, str2]).into_condition());
```
Though it didn't work, because `?` doesn't get translated by seaorm/sqlx when inside a square bracket.

**Reply by billy1624** [reply]
For PostgreSQL, it use `$N` (ordered parameters, aka `$1`, `$2`, `$3` and so on) instead of `?`.

**Reply by tasn** [reply]
`cust_with_values` (or maybe the query?) automatically transforms `?` to `$N` in other cases (that's why I had to do `??|` instead of `?|` for example), though not here. Though yeah, I guess I can just manually use `$N` directly here hoping it doesn't get escaped.

**Reply by billy1624** [reply]
Yeah, go with `$N` :D
